File: examples/multi-language/scripts/python/transform.py
import json
import os
import yaml
from datetime import datetime, timezone

# Read input from previous step
input_path = "/input/python-analyze/enriched.json"
with open(input_path) as f:
    enriched = json.load(f)
print(f"Read enriched data from {input_path}")

# Create final summary
total_languages = sum(len(langs) for langs in enriched["languagesByCategory"].values())
best_category = max(enriched["categories"].items(), key=lambda x: x[1]["avgScore"])

summary = {
    "generatedBy": "python-transform",
    "timestamp": datetime.now(timezone.utc).isoformat(),
    "pipeline": "multi-lang-example",
    "stepsCompleted": ["node-analyze", "node-transform", "python-analyze", "python-transform"],
    "totalLanguages": total_languages,
    "bestCategory": {
        "name": best_category[0],
        "avgScore": best_category[1]["avgScore"],
        "rating": best_category[1]["rating"],
    },
    "allCategories": {
        cat: {
            "rating": info["rating"],
            "avgScore": info["avgScore"],
            "languages": enriched["languagesByCategory"].get(cat, []),
        }
        for cat, info in enriched["categories"].items()
    },
}

# Write both YAML and JSON outputs
yaml_path = os.path.join("/output", "summary.yaml")
json_path = os.path.join("/output", "summary.json")

with open(yaml_path, "w") as f:
    yaml.dump(summary, f, default_flow_style=False, sort_keys=False)
print(f"Summary YAML written to {yaml_path}")

with open(json_path, "w") as f:
    json.dump(summary, f, indent=2)
print(f"Summary JSON written to {json_path}")

print(f"Pipeline complete! Best category: {best_category[0]} ({best_category[1]['avgScore']} avg)")

File: examples/multi-language/scripts/nodejs/transform.js
const fs = require('fs')
const path = require('path')
const {sortBy, flatMap} = require('lodash')

// Read input from previous step
const inputPath = path.join('/input/node-analyze', 'analysis.json')
const analysis = JSON.parse(fs.readFileSync(inputPath, 'utf8'))
console.log(`Read analysis from ${inputPath}`)

// Transform: create a ranked report
const allLanguages = flatMap(
  Object.entries(analysis.analysis),
  ([category, data]) => data.languages.map((lang, i) => ({
    language: lang,
    category,
    rank: i + 1
  }))
)

const ranked = sortBy(allLanguages, 'language')

const report = {
  generatedBy: 'node-transform',
  timestamp: new Date().toISOString(),
  sourceStep: 'node-analyze',
  totalLanguages: analysis.totalLanguages,
  ranked,
  summary: Object.fromEntries(
    Object.entries(analysis.analysis).map(([cat, data]) => [
      cat,
      {count: data.count, avgScore: data.avgScore}
    ])
  )
}

const outputPath = path.join('/output', 'report.json')
fs.writeFileSync(outputPath, JSON.stringify(report, null, 2))
console.log(`Report written to ${outputPath}`)
console.log(`Total languages ranked: ${ranked.length}`)
